src/main/java/org/folio/rest/impl/PostOrderLineHelper.java

```diff
@@ -56,34 +86,179 @@
+    } catch (UnsupportedEncodingException e) {
+      logger.error(String.format("Error during query encoding: %s", e.getMessage()));
+      throw new CompletionException(e);
+    }
+
+    return handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger)
+      .thenCompose(instances -> {
+        if (instances.getJsonArray("instances").size() > 0) {
+          return completedFuture(instances.getJsonArray("instances").getJsonObject(0).getString("id"));
+        }
+        return createInstanceRecord(compPOL, productTypesMap);
       });
-    return future;
+  }
+
+  /**
+   * Creates Instance Record in Inventory and returns its Id.
+   *
+   * @param compPOL PO line to create Instance Record for
+   * @param productTypesMap product types Map used to build Instance Record json object
+   * @return id of newly created Instance Record
+   */
+  private CompletableFuture<String> createInstanceRecord(PoLine compPOL, Map<String, String> productTypesMap) {
+    JsonObject lookupObj = new JsonObject();
+    CompletableFuture<Void> instanceTypeFuture = getInstanceType(DEFAULT_INSTANCE_TYPE_CODE)
+      .thenAccept(lookupObj::mergeIn);
+    CompletableFuture<Void> statusFuture = getStatus(DEFAULT_STATUS_CODE)
+      .thenAccept(lookupObj::mergeIn);
+
+    return VertxCompletableFuture.allOf(ctx, instanceTypeFuture, statusFuture)
+      .thenApply(v -> buildInstanceRecordJsonObject(compPOL, productTypesMap, lookupObj))
+      .thenCompose(instanceRecJson -> {
+        try {
+          return httpClient.request(HttpMethod.POST, instanceRecJson.toBuffer(), "/inventory/instances", okapiHeaders)
+            .thenApply(response -> {
+              logger.debug("Validating received response");
+              if (!isSuccess(response.getCode())) {
+                throw new CompletionException(
+                  new HttpException(response.getCode(), response.getError().getString("errorMessage")));
+              }
+              return response;
+            })
+            .thenApply(response -> {
+              String location = response.getHeaders().get(LOCATION_HEADER);
+              return location.substring(location.lastIndexOf('/'));
+            });
+        } catch (Exception e) {
+          throw new CompletionException(e);
+        }
+      });
+  }
+
+  private CompletableFuture<JsonObject> getInstanceType(String typeName) {
+    String endpoint = String.format("/instance-types?query=code=%s", typeName);
+    return handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger);
+  }
+
+  private CompletableFuture<JsonObject> getStatus(String statusCode) {
+    String endpoint = String.format("/instance-statuses?query=code=%s", statusCode);
+    return handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger);
+  }
+
+  private String buildProductIdQuery(ProductId productId, Map<String, String> productTypes) {
+    return String.format("(identifiers adj \"\\\"identifierTypeId\\\": \\\"%s\\\"\" " +
```

Comment:
I think for identifiers and types we should be using "==" instead of "adj".  We want exact matches in both cases